server/RdtClient.Service/Services/Torrents.cs

```diff
@@ -107,13 +107,51 @@ public async Task UpdateCategory(String hash, String? category)
         await torrentData.UpdateCategory(torrent.TorrentId, category);
     }
 
-    public async Task<Torrent> AddMagnetToDebridQueue(String magnetLink, Torrent torrent)
+    public static async Task<string> EnrichMagnetLink(string magnetLink)
     {
-        MagnetLink magnet;
+        var enrichment = Settings.Get.General.MagnetTrackerEnrichment;
+        if (enrichment == MagnetTrackerEnrichment.None) { return magnetLink; }
+
+        var url = $"https://github.com/ngosang/trackerslist/raw/refs/heads/master/{enrichment}.txt";
+        var newTrackers = (await new HttpClient().GetStringAsync(url)).Split('\n', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries).Distinct(StringComparer.OrdinalIgnoreCase).ToArray();
+
+        var originalUri = new Uri(magnetLink);
+        var decodedParams = HttpUtility.ParseQueryString(originalUri.Query);
+        var rawXt = decodedParams["xt"]!;
+        decodedParams.Remove("xt");
+
+        var existingTrackers = decodedParams.GetValues("tr") ?? Array.Empty<string>();
+        decodedParams.Remove("tr");
+        var allTrackers = existingTrackers.Concat(newTrackers).Distinct(StringComparer.OrdinalIgnoreCase);
+
+        var pieces = new List<string> { $"xt={rawXt}" };
+        var allKeys = decodedParams.AllKeys;
+        if (allKeys != null)
+        {
+            foreach (string? key in allKeys)
+            {
+                if (key == null) continue;
+                foreach (var val in decodedParams.GetValues(key)!)
+                {
+                    pieces.Add($"{key}={HttpUtility.UrlEncode(val)}");
+                }
+            }
+        }
+        foreach (var tr in allTrackers) { pieces.Add($"tr={HttpUtility.UrlEncode(tr)}"); }
+
+        var basePart = originalUri.GetLeftPart(UriPartial.Path);
+        var enriched = $"{basePart}?{string.Join("&", pieces)}";
+
+        return enriched;
```

Comment:
I don't love assembling the query string by hand. We can just use `.ToString()` on the return value of `HttpUtility.ParseQueryString()` to get the query params back as a string:

```suggestion
        var originalUri = new Uri(magnetLink);
        var queryParams = HttpUtility.ParseQueryString(originalUri.Query);
        
        foreach (var tr in newTrackers)
        {
            queryParams.Add("tr", tr);
        }

        return new UriBuilder(originalUri)
        {
            Query = queryParams.ToString()
        }.ToString();
```

Reply:
I considered appending all trackers to the list, but I had two issues. The URI can become too long for monotorrent when it goes to parse, so I broke it down to de-duplicate the trackers, and the other issue was overencoding of the URI from `.ToString()`, causing problems with monotorrent not recognising `xt`

Reply:
You're completely right; I should have tested more thoroughly before suggesting that.

Your solution seems like the most sensible way.